### PR TITLE
Add reasons to `skipif`

### PR DESCRIPTION
**Pull request by WilliamJamieson**
### Description

This PR is part of breaking apart #13754 into smaller less controversial parts.

In #13754 @astrofrog suggested that we should write all our `skipif` decorators using actual python code. This PR implements this for all `HAS_`<package name> based `skipif` decorators.

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Do the proposed changes actually accomplish desired goals?
- [x] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [x] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [x] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [x] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [x] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label. Codestyle issues can be fixed by the [bot](https://docs.astropy.org/en/latest/development/workflow/development_workflow.html#pre-commit).
- [x] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label. If this is a manual backport, use the `skip-changelog-checks` label unless special changelog handling is necessary.
- [x] Is this a big PR that makes a "What's new?" entry worthwhile and if so, is (1) a "what's new" entry included in this PR and (2) the "whatsnew-needed" label applied?
- [x] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [x] At the time of adding the milestone, if the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

### CODE REVIEW

**Comment by WilliamJamieson** [comment]
Can we enforce this with a pre-commit hook?

File: astropy/io/ascii/tests/test_connect.py [hunk 1]
```diff
@@ -1,4 +1,6 @@
 # Licensed under a 3-clause BSD style license - see LICENSE.rst
+import re
```
Comment (pllim) [inline astropy/io/ascii/tests/test_connect.py:2]: Why is this added?

Comment (WilliamJamieson) [reply astropy/io/ascii/tests/test_connect.py]: Apparently, autocomplete went wild when I made the change here. This was a mistake.

File: astropy/modeling/tests/test_quantities_fitting.py [hunk 2]
```diff
@@ -224,7 +224,7 @@ def test_compound_without_units(model, fitter):
 
 
 # FIXME: See https://github.com/astropy/astropy/issues/10675
-# @pytest.mark.skipif('not HAS_SCIPY')
+# @pytest.mark.skipif(not HAS_SCIPY, reason='requires scipy')
 @pytest.mark.skip(reason='Flaky and ill-conditioned')
```
Comment (pllim) [inline astropy/modeling/tests/test_quantities_fitting.py:228]: @WilliamJamieson , you don't want to fix this while you are at it? 😆

**Comment by WilliamJamieson** [comment]
> Can we enforce this with a pre-commit hook?

How I found all these fixes was adding this pre-commit hook:
```
      - id: skipif-reason
        name: check for reason in skipif decorators
        entry: '(?i)@(.*)skipif(?!(.|\n)*reason)'
        language: pygrep
        types: [python]
```
However, It fails for multi-line `skipif`.

**Comment by dhomeier** [comment]
> However, It fails for multi-line `skipif`.

Meaning it does not detect a `reason` in a continuation line? What is the `\n` in the regex doing?

**Comment by dhomeier** [comment]
Running that directly as `re.match` works for me on multi-line strings as well, sounds like the problem will rather be to get pre-commit to use the multi-line grep.
